staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -446,28 +452,110 @@ func (p *patcher) applyAdmission(ctx context.Context, patchedObject runtime.Obje
+		}
+		oldObjVersioned, err := convert(oldObj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert old object to proper version: %v", err)
+		}
+
+		managed, err := apply.DecodeObjectManagedFields(newObj)
+		if err != nil {
+			return nil, err
+		}
+
+		newObjTyped, err := typeConverter.ObjectToTyped(newObjVersioned)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create typed new object: %v", err)
+		}
+		oldObjTyped, err := typeConverter.ObjectToTyped(oldObjVersioned)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create typed old object: %v", err)
+		}
+
+		managed, err = updater.Update(oldObjTyped, newObjTyped, managed, manager)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update ManagedFields: %v", err)
+		}
+
+		if err := apply.EncodeObjectManagedFields(newObj, managed); err != nil {
+			return nil, err
+		}
+
+		return newObj, nil
+	}
+}
+
 // patchResource divides PatchResource for easier unit testing
 func (p *patcher) patchResource(ctx context.Context, scope RequestScope) (runtime.Object, bool, error) {
+	var (
+		converter apply.TypeConverter
+		updater   merge.Updater
+	)
+
+	modelCheck := func(models openapiproto.Models, manager string) (rest.TransformFunc, error) {
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServerSideApply) {
+			if scope.OpenAPIModels != nil {
+				var err error
+				converter, err = apply.NewTypeConverter(scope.OpenAPIModels)
+				if err != nil {
+					return noOpTransform, err
+				}
+
+				updater = merge.Updater{
+					Converter: apply.NewVersionConverter(converter, p.unsafeConvertor),
+				}
+				return makeManagedFieldsUpdater(converter, newConvertor(p.unsafeConvertor, p.kind.GroupVersion()), updater, manager), nil
+			}
+			return noOpTransform, fmt.Errorf("No OpenAPI for: %v", scope.Kind)
+		}
+		return noOpTransform, fmt.Errorf("server side apply feature gate disabled")
+	}
+
+	p.applyManagedFields = noOpTransform
 	p.namespace = request.NamespaceValue(ctx)
 	switch p.patchType {
 	case types.JSONPatchType, types.MergePatchType:
 		p.mechanism = &jsonPatcher{patcher: p}
+		p.applyManagedFields, _ = modelCheck(scope.OpenAPIModels, "jsonpatch")
 	case types.StrategicMergePatchType:
 		schemaReferenceObj, err := p.unsafeConvertor.ConvertToVersion(p.restPatcher.New(), p.kind.GroupVersion())
 		if err != nil {
 			return nil, false, err
 		}
 		p.mechanism = &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
-	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
+		p.applyManagedFields, _ = modelCheck(scope.OpenAPIModels, "smpatch")
+		// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
 	case types.ApplyPatchType:
-		p.mechanism = &applyPatcher{patcher: p, models: scope.OpenAPIModels}
+		if _, err := modelCheck(scope.OpenAPIModels, ""); err != nil {
+			klog.Warningf("%v", err)
+			return nil, false, nil
```

Comment:
you should probably return the error here, rather than warn, since it is mandatory here?

Reply:
Also, you can be here if the flag is not enabled